### PR TITLE
Better snakemake cluster submission support

### PR DESCRIPTION
**Pull request by AroneyS**
Add rule to localrules if simple, otherwise add resources with estimated cpu/mem/time.

Not sure about all the long-reads assemblies resources.

- [x] Add multiplier for resources with retry (if submitted job fails)
- [x] Add proper arguments to support snakemake args
  - [x] e.g. `--profile qsub --retries 3`, where profile points to `~/.config/snakemake/qsub/config.yaml` which contains cluster, cluster-status, jobs, cluster-cancel and other snakemake arguments
- [x] Add logs for each rule (stdout/stderr captured by cluster software)

### CODE REVIEW

**Review by rhysnewell — DISMISSED** [review]

**Comment by wwood** [comment]
localrules - how did I miss that before. Great.

**Comment by rhysnewell** [comment]
Let me know when you're ready for this to be reviewed

**Comment by AroneyS** [comment]
@rhysnewell @wwood Found a bug in snakemake where it solves the dag and then exits with no error. Only happened with >1 binner with cluster-submission. After much debugging, it turns out that its because the refinery rules don't have the `group:` keyword. So, since cluster submission tries to combine groups, it couldn't find anything to submit.

Anyway, solution is to either add the refinery rules to `group: 'binning'`, or remove groups entirely. I'm leaning towards the latter (or to replace them with specific grouped rules) to make the submitted jobs smaller. Was there a reason that the groups were added in the first place? I see they are split into binning/assembly/annotation/isolate...

**Comment by wwood** [comment]
Who wrote them in according to git blame? I thought they were only for cluster/cloud submission, so probably best to delete, I think, but maybe there's a reason I don't appreciate.

**Comment by rhysnewell** [comment]
I would have added in those groups ages ago in an attempt to organise the rules a bit better, but if they are just causing issues then just remove them

**Comment by rhysnewell** [comment]
I just had a quick look, and stdout is being written to the log file for both the run_checkm and refine_rosella rules. All logging info for rosella comes from stderr, so you'd need to pipe stderr to the log file. Stdout should generally remain empty for a rosella run. I could be incorrect though, just from a glance

**Comment by AroneyS** [comment]
The stdout and stderr should both be going to log files since stderr is redirected to STDOUT. I did that since some of the tools aren't well-behaved...

```python
subprocess.run(
    f"checkm2 predict -i {bin_folder}/ -x {bin_ext} -o {output_folder} -t {threads} --force".split(),
    env=os.environ,
    stdout=logf,
    stderr=subprocess.STDOUT
    )
```

**Comment by rhysnewell** [comment]
Ah gotcha, hmm. I guess you'd have to check firstly that the log output isn't just being piped somewhere, and secondly that the refinement is even running? Like it might be getting skipped in your tests due to lack of bins or something

**Comment by AroneyS** [comment]
Hah, yeh of course. The input CheckM files are empty. I'll log that refinement is being skipped.

**Comment by AroneyS** [comment]
@rhysnewell @wwood I think its ready. Good luck reading +1191/-676 lines, at least most of them are just removing groups and adding resources/logs.

File: aviary/aviary.py [hunk 1]
```diff
@@ -195,6 +195,23 @@ def main():
         default=" "
     )
 
+    base_group.add_argument(
+        '--snakemake-profile',
+        help='Snakemake profile (see https://snakemake.readthedocs.io/en/stable/executing/cli.html#profiles)\n'
+             'Create profile as `~/.config/snakemake/[CLUSTER_PROFILE]/config.yaml`. \n'
```
Comment (wwood) [inline aviary/aviary.py:201]: Unclear whether you mean the user or aviary does the creating. I think an example yml in the doco would go a long way here.

**Review by wwood — COMMENTED** [review]
Hi Sam, thanks for all this work. I didn't have time to go through each and every line, but the ones I looked at were good.

Happy to see there's further testing going on too.

I've only really requested a bit more documentation, I think a short example on using the cluster profile would work well enough.

Beyond this, I think it would be worth gathering a list of advantages of Aviary over other workflows, and listing those on the README. Can you please make a start on that, maybe in another PR?

Ta

**Review by rhysnewell — APPROVED** [review]

File: aviary/aviary.py (hunk 1 above)
Comment (AroneyS) [reply aviary/aviary.py:201]: I added some stuff under Advanced Usage. Is that what you wanted?

**Comment by wwood** [comment]
Sounds good, but maybe add a subtitle rather than just advanced usage.

Also, is this still true? From that same doc. Would be worth adding a few words to say that it is only sorta true e.g. reruns in a cluster mode are requeued asking for increasingly more RAM

```
When performing assembly, users are required to estimate how much RAM they will need to use via -m, --max-memory, --max_memory
```

**Comment by AroneyS** [comment]
> 

The memory arg is used as a cap on job submission memory. So the job memory will increase from a set amount with each retry but stop at the cap.

**Pull request merged**